### PR TITLE
Jekyll search functionality

### PR DESCRIPTION
**Pull request by caquino**
This pull request adds search functionality to the template using [lunrjs](https://lunrjs.com/)

All changes are guarded inside of if clauses, so unless `site_search` is set to true on `_config.yml` it should be noop.

### CODE REVIEW

File: assets/js/lunr-feed.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+---
+---
+{% if site.site_search %}
```
Comment (caquino) [inline assets/js/lunr-feed.js:4]: @crisostomon I've encapsulated the search database with an if clause to not impact the jekyll page generation process.

File: assets/js/lunr-feed.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    "id": {{ counter }},
+    "url": "{{ site.url }}{{ page.url }}",
+    "title": "{{ page.title }}",
+    "body": "{{ page.content | markdownify | replace: '.', '. ' | replace: '</h2>', ': ' | replace: '</h3>', ': ' | replace: '</h4>', ': ' | replace: '</p>', ' ' | strip_html | strip_newlines | replace: '  ', ' ' | replace: '"', ' ' }}"{% assign counter = counter | plus: 1 %}
+    }, {% endfor %}{% for page in site.posts %}{
+    "id": {{ counter }},
+    "url": "{{ site.url }}{{ page.url }}",
+    "title": "{{ page.title }}",
+    "body": "{{ page.date | date: "%Y/%m/%d" }} - {{ page.content | markdownify | replace: '.', '. ' | replace: '</h2>', ': ' | replace: '</h3>', ': ' | replace: '</h4>', ': ' | replace: '</p>', ' ' | strip_html | strip_newlines | replace: '  ', ' ' | replace: '"', ' ' }}"{% assign counter = counter | plus: 1 %}
+    }{% if forloop.last %}{% else %}, {% endif %}{% endfor %}];
```
Comment (robtuley) [inline assets/js/lunr-feed.js]: This seems a little awkward. I [googled this](https://talk.jekyllrb.com/t/loop-through-all-pages-of-the-website/6880/2), which suggests this sort of thing should work:

```
{% assign pages = site.pages | where_exp: "item", "item.title" %}
{% assign posts = site.posts %}
{% assign noPlugin = site.without-plugin %}
{% assign allPages = pages | concat: posts | concat: noPlugin %}
{% for page in allPages %}
    ... then write your logic once ...?
{% endfor %}
```

File: assets/js/lunr-feed.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    this.field('title')
+    this.field('body')
+
+    documents.forEach(function (doc) {
+        this.add(doc)
+    }, this)
+});
+
+function initSearch() {
+    console.log("init search...")
```
Comment (robtuley) [inline assets/js/lunr-feed.js]: is this intentional? I don't think we console log from anywhere else

**Review by robtuley — CHANGES_REQUESTED** [review]
Nice 🚀  -- just a couple of minor tweaks from me, but really nice neat solution.

File: assets/js/lunr-feed.js (hunk 3 above)
Comment (caquino) [reply assets/js/lunr-feed.js]: that is not intentional, I'm gonna remove it.

File: assets/js/lunr-feed.js (hunk 2 above)
Comment (robtuley) [reply assets/js/lunr-feed.js]: Plus random thought... wouldn't this produce invalid json if there were no posts? e.g. `{% if forloop.last %}{% else %}, {% endif %}{% endfor %}` only applies in the last of the 3 loops, which might be an empty loop etc?

Comment (caquino) [reply assets/js/lunr-feed.js]: is generates a valid empty json, I tested it. Well at least with the existing code.

Comment (caquino) [reply assets/js/lunr-feed.js]: I did rewrite this part to improve readability, it worked but it was too complex. Can you take a look on the new version?

File: _includes/header.html [hunk 4]
```diff
@@ -42,17 +42,14 @@
     <div class="flex items-center md:ml-12">
```
Comment (crisostomon) [inline _includes/header.html:42]: I don't think this will adjust in a responsive way when viewing on Mobile. Did you test the layout on Mobile view?

**Review by crisostomon — CHANGES_REQUESTED** [review]

**Comment by caquino** [comment]
@crisostomon just to avoid going blind and testing every single possible resolution, is it possible for us to check on google analytics which devices/resolutions we need to support?

**Comment by crisostomon** [comment]
> @crisostomon just to avoid going blind and testing every single possible resolution, is it possible for us to check on google analytics which devices/resolutions we need to support?

You can use the [tailwind breakpoints](https://tailwindcss.com/docs/responsive-design) so that it applies different behaviours when changing screen sizes and it's quite accessible to use Don't forget that tailwind uses a Mobile first approach.

**Comment by caquino** [comment]
> > @crisostomon just to avoid going blind and testing every single possible resolution, is it possible for us to check on google analytics which devices/resolutions we need to support?
> 
> You can use the [tailwind breakpoints](https://tailwindcss.com/docs/responsive-design) so that it applies different behaviours when changing screen sizes and it's quite accessible to use Don't forget that tailwind uses a Mobile first approach.

I understand that you can do responsive with anything, it just a matter of effort, but I was curious about which resolutions I need to test.

Actually this got me curious, should any PR sent to this repository be responsive and tested in all possible resolutions indepent of usage?

If so maybe it will be simpler for me to override the needed files on the gbp-site repo and just ignore what is not needed by our users, as I have no experience with tailwind/responsive options for it.

One important thing, I don't see how we will be able to fit the search counter and search box on the same line on a lower resolution screen, so either we need to remake the functionality and remove the search result counter, or we don't support resolutions smaller than a certain size.

**Comment by crisostomon** [comment]
> > > @crisostomon just to avoid going blind and testing every single possible resolution, is it possible for us to check on google analytics which devices/resolutions we need to support?
> > 
> > 
> > You can use the [tailwind breakpoints](https://tailwindcss.com/docs/responsive-design) so that it applies different behaviours when changing screen sizes and it's quite accessible to use Don't forget that tailwind uses a Mobile first approach.
> 
> I understand that you can do responsive with anything, it just a matter of effort, but I was curious about which resolutions I need to test.
> 
> Actually this got me curious, should any PR sent to this repository be responsive and tested in all possible resolutions indepent of usage?
> 
> If so maybe it will be simpler for me to override the needed files on the gbp-site repo and just ignore what is not needed by our users, as I have no experience with tailwind/responsive options for it.
> 
> One important thing, I don't see how we will be able to fit the search counter and search box on the same line on a lower resolution screen, so either we need to remake the functionality and remove the search result counter, or we don't support resolutions smaller than a certain size.

By default, we test the responsiveness of the layout base on the defaults of the framework used, that in this case is Tailwind. Those are defined [here](https://tailwindcss.com/docs/screens).

In case of the fitting of elements in the search box, usually, I would hide the counter because is only extra information
and not mandatory for the search to work.

I'll commit some changes related to this to your repo.

**Review by crisostomon — APPROVED** [review]

**Review by robtuley — APPROVED** [review]

**Comment by caquino** [comment]
Thanks for the help @crisostomon @robtuley !

**Pull request merged**